sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java

```diff
@@ -487,22 +481,41 @@ private HttpPipeline setupPipeline(HttpClient httpClient) {
                 return getPublicClientApplication(false)
                     .acquireTokenSilently(parametersBuilder.build());
             } catch (MalformedURLException e) {
-                throw logger.logExceptionAsError(Exceptions.propagate(e));
+                return getFailedCompletableFuture(logger.logExceptionAsError(Exceptions.propagate(e)));
             }
         }).map(ar -> new MsalToken(ar, options))
-        .filter(t -> !t.isExpired())
-        .switchIfEmpty(Mono.fromFuture(() -> {
-            SilentParameters.SilentParametersBuilder forceParametersBuilder = SilentParameters.builder(
-                new HashSet<>(request.getScopes())).forceRefresh(true);
-            if (account != null) {
-                forceParametersBuilder = forceParametersBuilder.account(account);
-            }
+            .filter(t -> !t.isExpired())
+            .switchIfEmpty(Mono.fromFuture(() -> {
+                SilentParameters.SilentParametersBuilder forceParametersBuilder = SilentParameters.builder(
+                    new HashSet<>(request.getScopes())).forceRefresh(true);
+                if (account != null) {
+                    forceParametersBuilder = forceParametersBuilder.account(account);
+                }
+                try {
+                    return getPublicClientApplication(false).acquireTokenSilently(forceParametersBuilder.build());
+                } catch (MalformedURLException e) {
+                    return getFailedCompletableFuture(logger.logExceptionAsError(Exceptions.propagate(e)));
+                }
+            }).map(result -> new MsalToken(result, options)));
+    }
+
+    /**
+     * Asynchronously acquire a token from the currently logged in client.
+     *
+     * @param request the details of the token request
+     * @return a Publisher that emits an AccessToken
+     */
+    public Mono<AccessToken> authenticateWithConfidentialClientCache(TokenRequestContext request) {
+        return Mono.fromFuture(() -> {
+            SilentParameters.SilentParametersBuilder parametersBuilder = SilentParameters.builder(
+                new HashSet<>(request.getScopes()));
             try {
-                return getPublicClientApplication(false).acquireTokenSilently(forceParametersBuilder.build());
+                return getConfidentialClientApplication().acquireTokenSilently(parametersBuilder.build());
             } catch (MalformedURLException e) {
-                throw logger.logExceptionAsError(Exceptions.propagate(e));
+                return getFailedCompletableFuture(logger.logExceptionAsError(Exceptions.propagate(e)));
             }
-        }).map(result -> new MsalToken(result, options))));
+        }).map(ar -> (AccessToken) new MsalToken(ar, options))
+            .filter(t -> !t.isExpired());
```

Comment:
Do we expect MSAL to return an expired token ?

Reply:
No. But a non-expired IAuthenticationResult could be an expired AccessToken because of the offset. In my token refresh offset re-work, this line will be changed to `.filter(t -> OffsetDateTime.now().isBefore(t.getExpiresAt().minus(identityClientOptions.getTokenRefreshOffset())));` But for now this may look a little confusing.